src/metabase/util.cljc

```diff
@@ -841,20 +841,38 @@
                          {:kvs kvs})))
        ret))))
 
-(defn classify-changes
+(defn row-diff
   "Given 2 lists of seq maps of changes, where each map an has an `id` key,
   return a map of 3 keys: `:to-create`, `:to-update`, `:to-delete`.
 
   Where:
-  :to-create is a list of maps that ids in `new-items`
-  :to-update is a list of maps that has ids in both `current-items` and `new-items`
-  :to delete is a list of maps that has ids only in `current-items`"
-  [current-items new-items]
-  (let [[delete-ids create-ids update-ids] (diff (set (map :id current-items))
-                                                 (set (map :id new-items)))]
-    {:to-create (when (seq create-ids) (filter #(create-ids (:id %)) new-items))
-     :to-delete (when (seq delete-ids) (filter #(delete-ids (:id %)) current-items))
-     :to-update (when (seq update-ids) (filter #(update-ids (:id %)) new-items))}))
+  - `:to-create` is a list of maps that ids in `new-rows`
+  - `:to-delete` is a list of maps that has ids only in `current-rows`
+  - `:to-skip`   is a list of identical maps that has ids in both lists
+  - `:to-update` is a list of different maps that has ids in both lists
+
+  Optional arguments:
+  - `id-fn` - function to get row-matching identifiers
+  - `to-compare` - function to get rows into a comparable state
+  "
+  [current-rows new-rows & {:keys [id-fn to-compare]
+                            :or   {id-fn   :id
+                                   to-compare identity}}]
+  (let [[delete-ids
+         create-ids
+         update-ids]     (diff (set (map id-fn current-rows))
+                               (set (map id-fn new-rows)))
+        known-map        (m/index-by id-fn current-rows)
+        {to-update false
```

Comment:
I still find it's amazing that you can destructure by the value of a map.

Reply:
Yeah it's a fantastic hack with group-by really :) it's like a continuous `split-by`.